syntax = "proto3";

package core.pb;
import "proto/renderer.proto";


message EditorPipeline
{
    Pipeline pipeline = 1;

    string vertex_shader_path = 8;// Used by the editor
    string fragment_shader_path = 9;// Used by the editor
    string compute_shader_path = 10;// Used by the editor
    string geometry_shader_path = 11;// Used by the editor
}


message EditorMaterialTexture
{
    MaterialTexture material_texture = 1;
    //Used only by editor
    string texture_name = 3;
    TextureType texture_type = 4;
}

message EditorMaterial
{
    repeated EditorMaterialTexture textures = 1;
    Material material = 2;

    string pipeline_path = 5; //used for the editor
}

message EditorMesh
{
    Mesh mesh = 1;

    string model_path = 2; //used by editor
}

message EditorModelDrawCommand
{
	repeated string material_paths = 1;
	repeated string draw_command_paths = 2;
	string pipeline_path = 3;
}

message EditorModel
{
	Model model = 1;
    repeated EditorModelDrawCommand draw_commands = 5; // used by the editor
    repeated string mtl_paths = 6;
}

message EditorDrawCommand
{
    DrawCommand draw_command = 1;
    //for the editor
    string material_path = 7;
    string mesh_path = 8;
}


message EditorSubPass
{
    SubPass subPass = 1;
    //Used by editor
    repeated string command_paths = 3;
    string framebuffer_path = 5;

}

message EditorRenderPass
{
    repeated EditorSubPass sub_passes = 1;
}

message EditorScene
{
    string name = 1;
    repeated Shader shaders = 2;
    repeated Pipeline pipelines = 3;
    RenderPass render_pass = 4;
    repeated Mesh meshes = 5;
    repeated Texture textures = 6;
    repeated Material materials = 7;
    repeated PySystem py_systems = 8;
    repeated string model_paths = 11;
    repeated FrameBuffer framebuffers = 12;
    repeated string cubemap_textures_side_paths = 13;
    //used by editor
    string render_pass_path = 9;
    repeated string py_system_paths = 10;
}